### PR TITLE
New: expose menu data as serializable objects

### PR DESCRIPTION
**Pull request by nytai**
exposes a method on the menu class to output the menu as list of dictionaries that can be converted to json.

### CODE REVIEW

File: flask_appbuilder/menu.py [hunk 1]
```diff
@@ -41,11 +41,12 @@ def __repr__(self):
 class Menu(object):
     menu = None
 
-    def __init__(self, reverse=True, extra_classes=""):
+    def __init__(self, reverse=True, extra_classes="", security_manager=None):
         self.menu = []
         if reverse:
             extra_classes = extra_classes + "navbar-inverse"
         self.extra_classes = extra_classes
+        self.security_manager = security_manager
```
Comment (dpgaspar) [inline flask_appbuilder/menu.py]: not needed, if we are calling get_data on a request context we can access the security manager with `current_app.appbuilder.sm`

File: flask_appbuilder/tests/test_menu.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        self.app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
+        logging.basicConfig(level=logging.ERROR)
+
+        self.db = SQLA(self.app)
+        self.appbuilder = AppBuilder(self.app, self.db.session)
+
+        sess = PSSession()
+
+        class PSView(ModelView):
+            datamodel = GenericInterface(PSModel, sess)
```
Comment (dpgaspar) [inline flask_appbuilder/tests/test_menu.py]: This can be removed

File: flask_appbuilder/tests/test_menu.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    def test_menu_api(self):
+        """
+            REST Api: Test menu data
+        """
+        uri = '/api/v1/menu/data'
+        client = self.app.test_client()
+        rv = client.get(uri)
+        eq_(rv.status_code, 200)
+        data = rv.data.decode('utf-8')
+        eq_(data, '{"data":[]}\n')
```
Comment (dpgaspar) [inline flask_appbuilder/tests/test_menu.py]: let's use `assertEqual` I'm changing everything on another PR

File: flask_appbuilder/menu.py [hunk 4]
```diff
@@ -123,3 +151,54 @@ def add_separator(self, category=""):
             raise Exception(
                 "Menu separator does not have correct category {}".format(category)
             )
+
+
+class MenuApi(BaseApi):
+    resource_name = "menu"
+
+    @expose('/data', methods=["GET"])
```
Comment (dpgaspar) [inline flask_appbuilder/menu.py]: I personally prefer just GET `/api/v1/menu`

File: flask_appbuilder/menu.py [hunk 5]
```diff
@@ -123,3 +151,54 @@ def add_separator(self, category=""):
...
+                                  type: string
+                                childs:
+                                  $ref: \
+        '#/paths/menu/data/get/reponses/200/content/application/json/schema/properties/data'
+                              required:
+                                - name
+                                - label
+                                - icon
+        """
+        return self.response(200, data=current_app.appbuilder.menu.get_data())
```
Comment (dpgaspar) [inline flask_appbuilder/menu.py]: Currently FAB uses `result` key instead of `data`

File: flask_appbuilder/menu.py [hunk 6]
```diff
@@ -123,3 +151,54 @@ def add_separator(self, category=""):
...
+                                  type: string
+                                icon:
+                                  type: string
+                                url:
+                                  type: string
+                                childs:
+                                  $ref: \
+        '#/paths/menu/data/get/reponses/200/content/application/json/schema/properties/result'
+                              required:
+                                - name
```
Comment (dpgaspar) [inline flask_appbuilder/menu.py]: `name` and `url` are required. `label`, `icon` are optional

File: flask_appbuilder/menu.py [hunk 7]
```diff
@@ -123,3 +151,54 @@ def add_separator(self, category=""):
             raise Exception(
                 "Menu separator does not have correct category {}".format(category)
             )
+
+
+class MenuApi(BaseApi):
+    resource_name = "menu"
+
+    @expose('/', methods=["GET"])
```
Comment (dpgaspar) [inline flask_appbuilder/menu.py]: public endpoint, is this intentional? it really does not return any data when no user is authenticated

File: flask_appbuilder/menu.py [hunk 8]
```diff
@@ -123,3 +151,54 @@ def add_separator(self, category=""):
...
+                                - name
+                                - label
+                                - icon
+        """
+        return self.response(200, result=current_app.appbuilder.menu.get_data())
+
+
+class MenuApiManager(BaseManager):
+    def register_views(self):
+        self.appbuilder.add_api(MenuApi)
```
Comment (dpgaspar) [inline flask_appbuilder/menu.py]: Let's create a new option key for this, take a look at: https://github.com/dpgaspar/Flask-AppBuilder/blob/master/flask_appbuilder/security/manager.py#L581

File: flask_appbuilder/menu.py (hunk 7 above)
Comment (nytai) [reply flask_appbuilder/menu.py]: Correct, it returns an empty array, currently. However if a user decides to expose a menu for the public role we'd still want to serve that [limited] menu, right?

File: flask_appbuilder/menu.py (hunk 8 above)
Comment (nytai) [reply flask_appbuilder/menu.py]: You mean expose a config to enable/disable this endpoint? something like `FAB_EXPOSE_MENU_API`?

File: flask_appbuilder/menu.py (hunk 7 above)
Comment (dpgaspar) [reply flask_appbuilder/menu.py]: Right, yet we do have a public role for this. It's actually a decision between exposing an unauthenticated EP that anonymously return an empty list, or exposing an authenticated EP that returns 401

File: flask_appbuilder/menu.py (hunk 8 above)
Comment (dpgaspar) [reply flask_appbuilder/menu.py]: yep

**Pull request merged**